scipy/sparse/csgraph/_shortest_path.pyx

```diff
@@ -166,12 +166,17 @@ def shortest_path(csgraph, method='auto',
         issparse = isspmatrix(csgraph)
         if issparse:
             Nk = csgraph.nnz
+            edges = csgraph.data
```

Comment:
This will fail for certain sparse matrix formats. CSR, CSC, and COO should work, but DIA, LIL, BSR, and DOK are likely to cause trouble. I'd add a test first to confirm the bug, then add a conditional `csgraph.tocoo()` conversion before extracting edges.

Reply:
While I was writing a test for this bug I found that `floyd_warshall` throws ValueError when graph is passed as sparse matrix other than LIL, CSR or CSC.

Here is the code that produces this error:

```
G_dense = np.array([[0, 2, 1],
                    [0, -1, 0],
                    [0, 0, 0]])
G_sparse = coo_matrix(G_dense)
shortest_path(G_sparse)
```
Traceback:
```
  File "_shortest_path.pyx", line 189, in scipy.sparse.csgraph._shortest_path.shortest_path
  File "_shortest_path.pyx", line 298, in scipy.sparse.csgraph._shortest_path.floyd_warshall
  File "/home/piurko/TCS/PZ/scipy/scipy/sparse/csgraph/_validation.py", line 29, in validate_graph
    csgraph = csgraph_to_dense(csgraph, null_value=null_value_out)
  File "_tools.pyx", line 320, in scipy.sparse.csgraph._tools.csgraph_to_dense
ValueError: csgraph must be lil, csr, or csc format
```
For this reason I'm not sure what to do with the test you asked for. Should I just run `shortest_path` function with some graph (for which other algorithm than FW will be selected) in every sparse matrix format?

Reply:
Oh, I didn't realize we had that restriction in place. Seems a bit too strict, but that's an issue for another time.

It should be sufficient to test with a LIL-format matrix.